# Video Quizzes
This notebook is for the computational/coding questions from lecture videos.

In [1]:
import numpy as np
import pandas as pd
import pprint
pp = pprint.PrettyPrinter(depth=6)
import random as rn

## Module 3 Quizzes

### Quiz 3.7 Constrained Optimization

In [3]:
np.argmin(np.square([-2,-1,1,2]))

1

## Module 4 Quizzes
### Quiz 4.8 Calculate MSE

In [87]:
w = np.array([1,2]) # [0,1,2] where 0 is bias
x = np.array([[1,2],[2,3]])
y = np.array([6,7])
n = x.shape[0]
MSE = (1/n) * np.sum(np.square(np.dot(w,x) - y))
MSE

1.0

### Quiz 4.9 Gradient Step Update

In [48]:
X = np.array([1,2,3]) # augment [2,3] to [1,2,3]
y = 6
alpha = .1
W = np.array([1,0,1])
gradient= X.dot(X.dot(W) - y) # X*(yhat - y)
W = W - (gradient*alpha) # Gradient update
W

array([1.2, 0.4, 1.6])

## Module 6 Quizzes
### Quiz 6.2 Probability Basics

In [47]:
p_a = .4
p_b = .6
p_a_and_b = 0
p_a_or_b = p_a + p_b - p_a_and_b
print(f'P(AUB) == P(A) + P(B) {p_a + p_b == 1}')
print(f'P(AUB) == P(A) + P(B) - P(A & B) {p_a + p_b == p_a_or_b}')
print(f'P(AUB) == P(A|B) {p_a_and_b/p_b == 1}')
print(f'P(AUB) == P(A)*P(B) {p_a*p_b == 1}')

P(AUB) == P(A) + P(B) True
P(AUB) == P(A) + P(B) - P(A & B) True
P(AUB) == P(A|B) False
P(AUB) == P(A)*P(B) False


### Quiz 6.3 Disjoing probabilities

In [51]:
# Disjoint means P(A & B) == 0
p_a = .6
p_b = .7
pa_and_b = 0
print(f'P(A) + P(B) - 0 == { p_a+p_b-pa_and_b == 1}')

P(A) + P(B) - 0 == False


### Quiz 6.4 Confidence Intervals

In [91]:
# Confidence Intervals
p = 1 - .2
n = 1000
z = 2
# phat +/- z/2sqrt(n)
wayOne = np.round(p + z/(2*np.sqrt(n)),2)
# Standard error (1/n * .2 *.8)
wayTwo = np.round(p + 2 * np.sqrt((1/n) *.2 *.8),2)
print(wayOne)
print(wayTwo)

0.83
0.83


### Quiz 6.15 Text Classification and Probability

In [141]:
data = ['Baseball','Helmet','Fans','Strike',
       'Football','Helmet','Touchdown',
       'Hoops',
       'Motorcycle','Helmet','Light',
       'Helmet','Ski']
classes = ['Yes','Yes','Yes','No','Yes']

# Q1: Vocab Size (unique count of words in training set)
print(f'Vocab size: {len(set(data))}')

# Q2: P(Class==Sports)
sports = classes.count('Yes')/len(classes)
print(f'P(Sports == Class) : {sports}')

# Q3: Conditional Probabilties
classes = ['Yes','Yes','Yes','Yes',
          'Yes','Yes','Yes',
          'Yes',
          'No','No','No',
          'Yes','Yes']
dictt = {'words':data,'sports':classes}
df = pd.DataFrame(data=dictt)

# Numerator: Helment where helmet is present and class is sport
#df[(df['words'] == 'Helmet') & (df['sports'] == 'Yes')]
numerator = len(df[(df['words'] == 'Helmet') & (df['sports'] == 'Yes')])

# Denominator: Words where class is sport
#df[(df['sports'] == 'Yes')]
denominator = len(df[(df['sports'] == 'Yes')])

print(f'P(word==Helment|Class==Sports): {numerator/denominator}')

# Q4: 1 + smoothing
# Add vocab size to denominator and 1 to numerator

print(f'+1 smoothing: {(3+1)/(10+10)}')

Vocab size: 10
P(Sports == Class) : 0.8
P(word==Helment|Class==Sports): 0.3
+1 smoothing: 0.2


## Module 7 Quizzes

In [5]:
# Quiz 7.7
import numpy as np
g = lambda t: 1/(1 + np.exp(-t))
w = np.array([6,0,-1])

# dummy values for x
x = np.array([1,0,6])

# Remember that o(X*w.T + B) == o(t)
# get probability of .5 when x2 == 6
print(g(np.dot(w,x)))

# x2 < 6
x = np.array([1,0,5])

# get probability when x2 < 6, p > .5 (class 1)
print(g(np.dot(x,w)))

# x2 > 6
x = np.array([1,0,7])

# get probability when x2 > 6, p < .5 (class 2)
print(g(np.dot(x,w)))

print('As x2 increases, p is below .5 (class 0). When x2 decreases, p is above .5 (class 1)')

0.5
0.7310585786300049
0.2689414213699951
As x2 increases, p is below .5 (class 0). When x2 decreases, p is above .5 (class 1)


In [61]:
# Quiz 7.2 Normalized Distnace
# 3x + 4y - 10 = 0
# (0,0)

(3*0 + 4*0 + 10)/np.sqrt(np.sum(np.square(np.array([3,4]))))

2.0

### Quiz 7.8 Class Softmax Classifier

In [368]:
# Info from problem
X = np.array([5,6])
W = np.array([[2,3],[1,-5],[1,1]])
b = np.array([4,10,12])

# Probability function for multinomial
p = lambda t: np.exp(t) / np.sum(np.exp(t))

# Classes
classes = ['A','B','C']

# Calculate scores and probabilities
scores = np.dot(X,W.T) + b
probs = p(scores)

# Print out class that the point will be assigned to
print(f'Class assigned: {classes[np.argmax(probs)]}')

Class assigned: A


### Quiz 7.9 Negative Log Loss Classifier

In [372]:
# Info from problem (assume true target class is 0)
perpDist = np.array([1.21,-.55,.15])

# Since 1.21 shows the true target class for 0
# The log loss calculation is only concerned with
# that distance

-np.log(np.exp(perpDist[0])/np.sum(np.exp(perpDist)))

0.41772344946346773

### Quiz 7.11 CXE (Binomial)

In [377]:
# Info from problem
W = np.array([2,-1,4,1])
X = np.array([1,5,1,2])
y = 0
p = lambda t: 1/(1 + np.exp(-t))
prob = p(X.dot(W.T))
CXE = -(y*np.log(prob) + (1-y) * np.log(1-prob))
CXE

3.048587351573745

### Quiz 7.12 Gradient Update (Binomial LR)

In [54]:
# Info from problem
W = np.array([2,-1,4,1])
X = np.array([1,5,1,2])
y = 0
p = lambda t: 1/(1 + np.exp(-t))
prob = p(X.dot(W.T))
lr = .01

gradient = np.dot((prob-y),X) # If you had more nested x values then 2/X.shape[0]
print(f'Gradient: {gradient}')

W = W - (lr * gradient)
print(f'New weight vector: {np.round(W,2)}')

Gradient: [0.95257413 4.76287063 0.95257413 1.90514825]
New weight vector: [ 1.99 -1.05  3.99  0.98]


# Fun with Modules

## Module 7

In [52]:
# Binomial probability function for logit
f = lambda t: 1/(1 + np.exp(-t))

# Create probabilities with neg t's
neg_p = [f((i*-1)/10) for i in range(0,11)]

# Create probabilities with pos t's
pos_p = [f(i/10) for i in range(0,11)]

# Quick analysis between t and p
print(f'When t == 0, then p is: {pos_p[0]}')
print('When t < 0')
pp.pprint(neg_p[1:10])
print('When t > 0')
pp.pprint(pos_p[1:10])

When t == 0, then p is: 0.5
When t < 0
[0.47502081252106,
 0.45016600268752216,
 0.425557483188341,
 0.401312339887548,
 0.3775406687981454,
 0.35434369377420455,
 0.3318122278318339,
 0.31002551887238755,
 0.289050497374996]
When t > 0
[0.52497918747894,
 0.549833997312478,
 0.574442516811659,
 0.598687660112452,
 0.6224593312018546,
 0.6456563062257954,
 0.6681877721681662,
 0.6899744811276125,
 0.7109495026250039]


In [33]:
# Playing around with multinomial regression
rn.seed(42)
X = np.array([rn.randint(-10,97) for i in range(5)])
W = np.array([(rn.randint(0,4),rn.randint(10,10),
               rn.randint(9,18),rn.randint(2,4),
               rn.randint(9,10)) for i in range(5)])
y = np.array([rn.randint(0,5) for i in range(5)])

# Weighted distance
dists = X.dot(W.T)

# Calculate probabilities
p = lambda t: np.exp(t) / np.sum(np.exp(t))
probs = p(dists)

# chosen class/probability
yhat = y[np.argmax(probs)]

# CXE
cxe = -y*np.log(probs)-(1-y)*np.log(1-probs)
cxe

array([1.74001823e+02, 1.40009115e+01, 3.33000911e+02, 9.11466454e-04,
       1.48001823e+02])

# Coding from assignments
## Homework 2

In [ ]:
## set up parameters ######################################
n_neighbors_range = list(range(1,6))
p_range = list(range(1,4))
###########################################################

## Set-up pipe-line and gridsearch ########################
pipe = Pipeline([
    ('standardize', StandardScaler()),
    ('knn',knn_reg)
])

# ORDER: pipe, parms, just tag the rest
knn_reg_gs = GridSearchCV(pipe,parameters,
                          cv=3,n_jobs=-1,
                          scoring="neg_mean_squared_error",
                          refit=True,verbose=1,n_jobs=-1)
############################################################

## Homework 5

In [ ]:
## Batch GD Update ############
def step(x, eta):
    update = eta * ((4*x) + 4)
    return x - update
step(x=2, eta=.1)
###############################

## Sochastic GD Update ##################################
X= [1,2,3]
W = [1, 0, 1]
y=[6]
alpha= 0.1
gradient= alpha*((np.array(X).dot(np.array(W)) - y) * X)
newW= W - gradient
print(newW)
#########################################################

## LG GD Update################# 
X=np.array([1, 5, 1, 2])
w=np.array([2,-1, 4, 8])
eta = .01
y = 6
w = w-eta*X.dot((X.dot(w) - y))
print(np.round(w,3))
################################

## Create a pipeline###################
pipe = Pipeline([
        ('scaler',StandardScaler()),
        ('lineareg',LinearRegression())
    ])
#######################################

## Calculate the STD Results########################################
table.iloc[s,(c*2)+1] = np.round(np.std(results[count]), decimals=3)
####################################################################

## Selecting K-best features #####################
Kbest = SelectKBest(f_regression,k=5)
X_train_sel = Kbest.fit_transform(X_train,y_train)
##################################################

## Calculate MAPE####################################################
mape = np.round(mean_absolute_percentage_error(y_val, y_val_pred), 3)
#####################################################################

## Separate x,y data and perform split training on training data ####
X = train.drop(['count','registered','casual'],axis=1)
y_reg, y_cas = train['registered'], train['casual']
#####################################################################

## Fit select-best features #########################################
Kbest_reg = Kbest_REG.fit(X_train,y_train_reg)
Kbest_cas = Kbest_CAS.fit(X_train,y_train_cas)
#####################################################################

## Calculate rsme ###################################################
rmse = np.round(np.sqrt(mean_squared_error(y_val_reg,y_pred_reg)), 3)
#####################################################################

# Transform data to log #
y = np.log(y + 1)
#########################

# Set-up sequential background selection #######
lr = LinearRegression()
sbs = SBS(lr,1,scoring=explained_variance_score)
################################################

# Pipeline with SBS ############################
pipe = Pipeline([
    ('scaler',StandardScaler()),
    ('sbs',SBSFeatureSelector(sbs.subsets_,5)),
    ('linreg',LinearRegression())
])
################################################

# Convert log back to non-log values ##################
y_pred = np.round((np.exp(y_pred_log) - 1).astype(int))
#######################################################

# Pipeline with multiple esimators #################################
pipe = Pipeline([
        ('scaler',StandardScaler()),
        (estimator)
])
gs = GridSearchCV(pipe,params,cv=5,scoring="neg_mean_squared_error")
####################################################################

# Closed form linear regression #################################
w = np.linalg.inv(X.transpose().dot(X)).dot(X.transpose()).dot(y)
#################################################################

## Calculate MSE #############
yhat = X.dot(w)
MSE = np.square(yhat-y).mean()
##############################

## Make-pipeline with SelectPercentile + Poly ########################
pipe = make_pipeline(
    StandardScaler(),
    SelectPercentile(score_func=f_regression),
    PolynomialFeatures(),
    Ridge())
param_grid = {'polynomialfeatures__degree': [1, 2, 3],
              'ridge__alpha': [0.001, 0.01, 0.1, 1, 10, 100], 
              'selectpercentile__percentile' : [10,20,50,65,70,80,90]
             }
#######################################################################

## Homework 7

In [ ]:
## Unnormalized perp dist #
scores = (np.dot(x,W) + b)
###########################

## Binary Log Loss ############################################################################
(-1/predicted.shape[0] * np.sum(actual * np.log(predicted) + (1-actual) * np.log(1-predicted)))
###############################################################################################

## Multinomial Log Loss (CXE) ##################
m = predictions.shape[0]
cxe = -1/m*np.sum(targets * np.log(predictions))
################################################

## GD Update (Binary) #######################
perpDist= np.dot(X,w.T)
p = 1 / (1 + np.exp(-perpDist)) # sigmoid
gradient = (1/X.shape[0]) * np.dot((p - y),X)
lr = 0.1
w = w - (gradient * lr)
#############################################

## Splitting up training/test data with sub_sample ######################################################################
X_train, _, y_train, _ = train_test_split(X_train, y_train, stratify=y_train, train_size=subsample_rate, random_state=42)
X_test, _, y_test, _ = train_test_split(X_test,y_test,stratify=y_test,train_size=subsample_rate,random_state=42) 
#########################################################################################################################

## Setting up k-fold and cross validation ##################################
kfold = KFold(n_splits=10, shuffle=True, random_state=7)

# set cv_results with scoring=scoring variable (which is set to 'accuracy')
cv_results = cross_val_score(model, X, Y, cv=kfold,scoring=scoring) 
############################################################################

## Pipeline with count vectors and tfidf #######
pipeline =  Pipeline([
    ('vect',CountVectorizer()),
    ('tfidf',TfidfTransformer()),
    ('clf',SGDClassifier(loss='log',max_iter=5))
])
################################################

## Set params, grids, and pipeline ###########################
scoring='precision_macro'
# select handful of parameters to explore
parameters = {'vect__ngram_range': ((1,1),(1,2)),
                 'tfidf__use_idf': (True, False),
                 'clf__alpha': (.001, .0001), 
                 'clf__penalty': ('l1', 'l2', 'elasticnet'), 
              # please explore different regularization terms 
              #('l1', 'l2', 'elasticnet'),
                 'clf__l1_ratio': (.15,  .50)}
grid_search = GridSearchCV(pipeline,parameters,
                           n_jobs=-1,verbose=1,
                           cv=3,scoring=scoring)  
###############################################################

## Logistic regression pipeline ########################
pipe = Pipeline([
     ('scaler',StandardScaler()),
     ('lasso',LogisticRegression(penalty="l1",C=param)),
    ])  
########################################################

# Labs
## Lab 2

In [ ]:
# splitting data ###################################################################################
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=.5, random_state=4)

# simple knn predictor ###################
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(sepal_train, y_train)
preds = knn.predict(sepal_test)

# homegrown knn ################################################################
# calculate euclidean distance
def metric_func(x1, x2):
    distance = np.linalg.norm(x1 - x2, ord=2, axis=1)
    return distance

examples = sepal_test[class_change]
nearest_indices = np.zeros(shape=(examples.shape[0], 5), dtype=np.int) - 1
nearest_distances = np.zeros(shape=(examples.shape[0], 5), dtype=np.float) - 1

# collect index and distance for 5 nearest neighbors
for i in range(examples.shape[0]):
    distances = metric_func(examples[i], sepal_train)
    index_order = np.argsort(distances)[:5]
    nearest_indices[i] = index_order
    nearest_distances[i] = distances[index_order]

# Creating KNN object and calculating a list of cross val scores #
knn = KNeighborsClassifier(n_neighbors=5)
scores = cross_val_score(knn, X, y, cv=10, scoring='accuracy')

# Also cross val score with logistic regression ###############
cross_val_score(logreg, X, y, cv=10, scoring='accuracy').mean()

# Sample with fold ######################################
kfold = KFold(n_splits=10, shuffle=True, random_state=42)
cross_val_score(knn, X_train, y_train, cv=kfold)

# Naming parameters ###############################################
from sklearn.model_selection import GridSearchCV
params = {'meow__n_neighbors': list(range(1,21)),
          'meow__weights': ['uniform', 'distance'],
          'meow__p': list(range(1,11))}
pipe = Pipeline([
    ('kitty',StandardScaler()),
    ('meow',KNeighborsClassifier())
])
knn_gs = GridSearchCV(pipe, params, n_jobs=-1, cv=5, verbose=1)
knn_gs.fit(X_train, y_train)

# HINT: use the parm names they give you, whatever your classifier is
# do param_key == classifierObjectName__restofparams

# Get best params #############
print(grid_search.best_params_)
print(grid_search.best_score_)
print(grid_search.best_estimator_)

# Print out keys ####################
print(grid_search.cv_results_.keys())

## Lab 3 Code

In [37]:
# Perform one step of gradient descent
f = lambda x : (4*x) + 4
x1 = 2 - .1 * f(2)
print(f'1 step gradient update: {x1}')

1 step gradient update: 0.7999999999999998


Find the minimum using GD:

| &emsp;&emsp;$ Iteration $&emsp;&emsp;| &emsp;&emsp; __$x_i$__ &emsp;&emsp;| &emsp;&emsp;&emsp;$f'(x_i)= 4*x_i +4$ &emsp;&emsp;&emsp;&emsp; | &emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp; $Step Update = \\x_i - \alpha * f'(x_i) $ &emsp;&emsp;&emsp;&emsp;&emsp;&emsp; |&emsp;&emsp; &emsp; $x_{i+i}$ &emsp;&emsp; &emsp; |
|:--------:|:-----------------:|:-----------------:|:-----------------:|:-----------------:|
|    1     | 2.0               |        $4*(2.0)+4$     |$2 - 0.1 * (4*(2.0)+4)$|      .7999        |
| 2 | .7999 | $4*(.7999)+4$ | $.7999 - 0.1 * (4*(.7999)+4)$ |  .0799   |
| 3 | .0799| $4*(.0799)+4$ |  $.0799 - 0.1 * (4*(.0799)+4)$ | -.3520 | 
|4 | -.3520| $4*(-.3520)+4$ | $-.3520 - 0.1 * (4*(-.3520)+4)$  | -.6112  |
| 5 | -.6112|$4*(-.6112)+4$| $-.6112 - 0.1 * (4*(-.6112)+4)$ | -.7667 |
| 6 | -.7667 | $4*(-.7667)+4$ | $-.7667 - 0.1 * (4*(-.7667)+4)$ | -.8600 |

In [51]:
# Homegrown GD
# is f(x) == 0, i.e., a root of f(x)
eta = .5
import pdb

def dx(f, x):
    return abs(0-f(x))

# find the zeros of the derivative function
# REPEAT find the zeros of the derivative function

eta = .5 # step size multiplier

#df is not used. WHY?
def gradientDescent(f, df, x0, e, print_res=True):
    delta = dx(f, x0)
    approximations = [x0]
    i=0
    while delta > e:
        #pdb.set_trace()
        x0 = x0 - eta * f(x0)
        if (i%10==0):
            approximations.append(x0)
        delta = dx(f, x0)
        #print(delta)
    if print_res:
        print ('Root is at: ', x0)
        print ('f(x) at root is: ', f(x0))
    return (x0, approximations)

# f is some sort of gradient function
f = lambda x: 3.5*x + 4
print(gradientDescent(f,0,2,.0000001))

# Newton's formula
from scipy.optimize import newton
from sklearn.utils.testing import assert_almost_equal

objectiveFunction = lambda x:x**2 - 4
f = lambda x : 2*x 
df = 2

#zero = newton(f, 1.0, fprime=df, maxiter=1000)
scipyRoot = newton(f, 2.7,  maxiter=100)  #vi scipy
(rootHomeGrown, trace) = gradientDescent(f, df,  2.7, 1e-5)

print(f'Newton: {rootHomeGrown}')
print(f'Trace:{trace}')

Root is at:  -1.1428571666423404
f(x) at root is:  -8.324819145144602e-08
(-1.1428571666423404, [2, -3.5, 0.625, -2.46875, -0.1484375, -1.888671875, -0.58349609375, -1.5623779296875, -0.828216552734375, -1.3788375854492188, -0.9658718109130859, -1.2755961418151855, -1.0433028936386108, -1.2175228297710419, -1.0868578776717186, -1.184856591746211, -1.1113575561903417, -1.1664818328572437, -1.1251386253570672, -1.1561460309821996, -1.1328904767633503, -1.1503321424274873, -1.1372508931793845, -1.1470618301154616, -1.1397036274134038, -1.1452222794399471, -1.1410832904200396, -1.1441875321849704, -1.1418593508612722, -1.143605486854046, -1.1422958848594655, -1.1432780863554008, -1.1425414352334495, -1.143093923574913, -1.1426795573188153, -1.1429903320108885, -1.1427572509918338, -1.1429320617561247, -1.1428009536829062, -1.1428992847378203, -1.1428255364466349, -1.1428808476650238, -1.1428393642512322, -1.1428704768115758, -1.142847142391318, -1.1428646432065115, -1.1428515175951164, -1.

## Lab 4 Code *Linear Regression*

In [63]:
# Full Gradient Descent
w = np.array([1,0])
x = np.array([[1,1],[2,1],[3,1],[4,1],[5,1]])
y = np.array([2,3,4,3,5])

# Compute prediction == yhat
yhat = x.dot(w)
gradient = np.dot(yhat-y,x)

# Compute rest of formula
w = w - .1 * 2/x.shape[0] * gradient
print(f'Weights after first Full GD/Batch update: {w}')

# Second step gradient
w = w - .1 * (2/x.shape[0]) * np.dot(x.dot(w) - y,x)
# Same as w - .1 * (2/5 * (w.dot(x.T) - y).dot(x))
print(f'Weights after second Full GD/Batch update: {w}')

# Stochastic Gradient Descent
w = np.array([1,0])
x = np.array([[1,1],[3,1],[5,1],[4,1],[2,1]])
y = np.array([2,4,5,3,3])

yhat = x[0].dot(w)
gradient = np.dot(yhat - y[0],x[0])
w = w - .1 * (2/x.shape[0]) * gradient
print(f'Weights after first Stochastic GD update: {w}')

yhat = x[1].dot(w)
gradient = np.dot(yhat - y[1],x[1])
w = w - .1 * (2/x.shape[0]) * gradient
print(f'Weights after second Stochastic GD update: {np.round(w,2)}')

# Compute total loss
w = np.array([1,0])
x = np.array([[1,1],[2,1],[3,1],[4,1],[5,1]])
y = np.array([2,3,4,3,5])

yhat = x.dot(w)
loss = 1/x.shape[0] * np.sum(np.square(yhat-y))
print(f'Total loss of function: {loss}')

Weights after first Full GD/Batch update: [1.08 0.08]
Weights after second Full GD/Batch update: [0.936 0.096]
Weights after first Stochastic GD update: [1.04 0.04]
Weights after second Stochastic GD update: [1.14 0.07]
Total loss of function: 0.8


In [173]:
import statsmodels.formula.api as sm
from sklearn.linear_model import LinearRegression
import pandas as pd
from sklearn.datasets import load_boston
import pandas as pd
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

# Feature Selection with KBest (just making a dataframe)
boston = load_boston()
X = pd.DataFrame(boston.data, columns=boston.feature_names)
y = boston.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

k_best = SelectKBest(f_regression)
feat_scores = k_best.fit_transform(X_train,y_train)

results = pd.DataFrame()
results["F Score"] = k_best.scores_
results["P Value"] = k_best.pvalues_
results["Support"] = k_best.get_support()
results["Features"] = X.columns

results.sort_values(by='F Score',ascending=False)

# Feature selection Kbest with pipeline
pipeline = Pipeline([
        ("std_scaler", StandardScaler()),
        ('feature_selection', SelectKBest(f_regression)),
        ('lineReg', LinearRegression())
    ])

pipeline.fit(X_train,y_train)
results = pipeline.predict(X_test)
df = pd.DataFrame({'Y Actual':y_test,'Y Predict':results})
df.head(10)

# Handling missing values

# drop nulls across rows
df.dropna(axis=0)

# drop nulls across cols
df.dropna(axis=1)

# drop nan rows
df.dropna(how='all')  

# drop in specific columns
df.dropna(subset=['C'])

# Use a mean to input missing values
from sklearn.preprocessing import Imputer

imr = Imputer(missing_values='NaN', strategy='mean', axis=0)
imr = imr.fit(df.values)
imputed_data = imr.transform(df.values)

Y Actual  Y Predict
0      23.6  27.531578
1      32.4  30.796233
2      13.6  14.931903
3      22.8  24.292355
4      16.1  19.976771
5      20.0  23.305321
6      17.8  18.400940
7      14.0  13.539295
8      19.6  23.677344
9      16.8  19.464453

## Lab 5 Code

In [131]:
# One sample lasso regression update
x = np.array([1,5,1,2])
w = np.array([2,-1,4,8])
y = 10
eta = .1
alpha = .3

gradient = np.dot(x.dot(w)-y,x)
lasso = np.append(0,alpha*np.sign(w[1:]))
w1 = w - eta * (gradient + lasso)
w1 = np.round(w1,2)
print(f'One sample lasso regression update: {w1}')

One sample lasso regression update: [ 1.3  -4.47  3.27  6.57]


In [147]:
X = np.array([1,5,1,2])
W = np.array([2,-1,4,8])
y = 10
lambda1 = .1
alpha = .3

Wn = W - alpha * np.transpose(X).dot(X.dot(W) - y)
Wn[1:] -= alpha * lambda1 * np.sign(W[1:])
np.round(Wn, 3)

array([ -0.1 , -11.47,   1.87,   3.77])

In [132]:
# One sample ridge regression update
x = np.array([1,5,1,2])
w = np.array([2,-1,4,8])
y = 10
eta = .1
alpha = .3

gradient = np.dot(x.dot(w)-y,x)
ridge = np.append(0,alpha*w[1:])
w2 = w - eta * (gradient + ridge)
w2 = np.round(w2,2)
print(f'One sample ridge regression update: {w2}')

One sample ridge regression update: [ 1.3  -4.47  3.18  6.36]


In [137]:
# One sample elastic net regression
x = np.array([1,5,1,2])
w = np.array([2,-1,4,8])
y = 10
eta = .1
alpha = .3
r = 1

gradient = np.dot(x.dot(w)-y,x)
lasso = np.append(0,alpha*r*np.sign(w[1:]))
ridge = np.append(0,alpha*((1-r)) * w[1:])
w3 = w - eta * (gradient + lasso + ridge)
w3 = np.round(w3,2)
print(f'Multi elastic gradient update: {w3}')

Multi elastic gradient update: [ 1.3  -4.47  3.27  6.57]


In [138]:
# Multiple lasso regression update
X = np.array([[1,5,1,2],[1,1,0,1],[1,4,3,2]])
w = np.array([2,-1,4,8])
y = np.array([10,1,3])
eta = .01
alpha = .1

gradient = np.dot(np.dot(X,w)-y,X)*(2/X.shape[0])
lasso = np.append(0,alpha*np.sign(w[1:]))
w1 = w - eta * (gradient + lasso)
w1 = np.round(w1,2)
print(f'Multi lasso gradient update: {w1}')

Multi lasso gradient update: [ 1.75 -1.9   3.49  7.55]


In [139]:
# Multi ridge regression update
X = np.array([[1,5,1,2],[1,1,0,1],[1,4,3,2]])
w = np.array([2,-1,4,8])
y = np.array([10,1,3])
eta = .01
alpha = .1

gradient = np.dot(np.dot(X,w)-y,X)*(2/X.shape[0])
ridge = np.append(0,alpha * w[1:])
w2 = w - eta * (gradient + ridge)
w2 = np.round(w2,2)
print(f'Multi ridge gradient update: {w2}')

Multi ridge gradient update: [ 1.75 -1.9   3.49  7.54]


In [140]:
# Multi elastic regression update
# When r == 0 then == ridge
# When r == 1 then == lasso

X = np.array([[1,5,1,2],[1,1,0,1],[1,4,3,2]])
w = np.array([2,-1,4,8])
y = np.array([10,1,3])
eta = .01
alpha = .1
r = 0

gradient = (2/X.shape[0]) * np.dot(X.dot(w) - y,X)
lasso = np.append(0,alpha*r*np.sign(w[1:]))
ridge = np.append(0,alpha*((1-r)) * w[1:])
w3 = w - eta * (gradient + lasso + ridge)
w3 = np.round(w3,2)
print(f'Multi elastic gradient update: {w3}')

Multi elastic gradient update: [ 1.75 -1.9   3.49  7.54]


# Lab 7

## Binomial  + Multinomial Logistic Regression

In [296]:
y = np.array([1,0])
w = np.array([-2,1,-1,2])
X = np.array([[1,2,1,2],[1,1,0,0]])
lr = 0.1

## Calculate probability for binary class
perpDist = X.dot(w) # AKA scores
p = 1/(1 + np.exp(-perpDist))
print(f'Binary LR Prob: {p}')

# Select classes
# if p < .5 then 0 else 1
print(f'Classes: {[1,0]}')

## Compute log loss (CXE) for binary class
CXE = (-1/X.shape[0]) * np.sum(y * np.log(p) + (1-y) * np.log(1-p))
print(f'Binary LR CXE: {CXE}')

## Compute binary gradient update
w  = w - lr * (1/X.shape[0]) * np.sum(np.dot((p-y),X))
print(f'Binary LR Gradient Update: {w}')

## Calculate probability for multinomial class
y = np.array([0,1,2])
w = np.array([-2,1,-1,2])
X = np.array([[1,2,1,0],[2,0,0,0],[2,2,2,1]])
lr = 0.1

perpDist = X.dot(w)
p = np.exp(perpDist)/np.sum(np.exp(perpDist))
print(f'Multinominal LR Prob: {p}')

# Select class
yhat = y[np.argmax(p)]
print(f'Selected class: {yhat}')

# Compute log loss (CXE) for multinomial class
CXE = (-1/X.shape[0]) * np.sum(y*np.log(p))
print(f'Multinominal LR CXE: {CXE}')

# Compute multinomial gradient update
w = w - lr * (1/X.shape[0]) * np.sum(np.dot(p-y,X))
print(f'Multinomial LR Gradient Update: {w}')

# Loss loss for class 0
print(f'Log loss for class 0: {-np.log(p[0])}')

# Loss loss for class 1
print(f'Log loss for class 1: {-np.log(p[1])}')

# Loss loss for class 2
print(f'Log loss for class 2: {-np.log(p[2])}')

Binary LR Prob: [0.95257413 0.26894142]
Classes: [1, 0]
Binary LR CXE: 0.18092451954598235
Binary LR Gradient Update: [-2.01266638  0.98733362 -1.01266638  1.98733362]
Multinominal LR Prob: [0.70538451 0.03511903 0.25949646]
Selected class: 0
Multinominal LR CXE: 2.0156788834348527
Multinomial LR Gradient Update: [-1.62360838  1.37639162 -0.62360838  2.37639162]
Log loss for class 0: 0.34901221676818633
Log loss for class 1: 3.3490122167681866
Log loss for class 2: 1.3490122167681862


# Module Summaries
## Module 2

**KNN**:
* Non-parametric method (doesn't conform to gaussian distributions + relies on continious data)
* Used for regression + classification
* Relies on K-closest neighbors/training examples in a feature space
* Case-based (retrieves stored data and adapts based on new case)
* Instance based (store all the training data)
* Supervised learning (training data has labels)
* Training phase == STORING DATA ONLY (labels and features)
* Classification phase: k and unlabeled vector assigned a label, based on most frequent k training samples with respect to test point

**KNN Regression**:

Minimize the following function:
    \begin{equation}
    MSE = \frac{1}{n}\Sigma (WX^i - y^i)^2
    \end{equation}
    y is the *real value*

**How to visualize it:**
1. Fit a normal regression line
2. Given an input x, draw a straight vertical line
3. Find k-nearest neighbors based on distance (use one of the three distance metrics as listed below)
4. Average the values of k-nearest neighbors to predict yhat

**Distances**
1. Euclidean
2. Manhattan
3. Minkowsi

**Standardize Data**
* Different features will have different scales, causing one to be a greater influence over the other. Scale the data to prevent this from happening
* Will occur when there's a mixture of numerical and categorical variables

\begin{equation}
X_i = \frac{X-Min}{Max-Min}
\end{equation}

**Decision Boundaries**
* Voronoi Tesselation is a way to visualize decision boundaries with KNN

**1-nearest neighbor classifier**
* As training size increases, 1 nearest neighbor results in minimum achievable error rate given the distrubtion of the data

**Complexity and generalization**
* Limit model complexity
* Split training and validation sets
* Efficient Implementations
    * Finding nearest neighbor is O(n)
    * Binary search tree reduces to O(log n)
    * Kd-tree (binary tree that sorts points on x then on y recursively)
    * Parallel Hardware: Const if O(n) processors and O(log n) time
    
**Pros of KNN**
* Fast training
* little bias
* Zero cost during learning unless
    * Memorize the data
    * Use kd trees

**Cons of KNN**
* Scaling
* Sensitivity to noise and non-important features
* Never used on image classificaton
    * Terrible performance at test time
    * distance metrics are not intuitive

**Image classification**
* Semantic gap: difference btwn descriptions of an object by different linguistic representations, like languages or symbols
    * Example: converting an image into an array of digits
* Challenges
    * Viewpoint variations
    * Illumination
    * Deformation
    * Occlusion
    * Background clutter
    * Variations
    * No way to really "code a cat"
* p/ord_p
    * 1 == manhattan  distance (l1)
    * 2 == euclidean_distance  (l2)
    * arbitrary p == minkoski_distance (l_p)
    * default is 2
* Extra
    * Training data size will cause classification speed to grow linearly
    
**Hyperparameter tuning**
* Split data into k-folds, one fold is validation data used to tune hyperparameters
* Average results from the fold

## Module 3

**What is optimization?**
* The process of finding solutions that give a maximum or miminum of a function
* Choses best avilable values of some objective function/domain
* Optimal solution == where objective function reaches max or min
* Global optimal is where no other feasible solutions with better values
* min -(f(x)) == max(f(x))
    * Optimization can mean minimization since max of a function is found by seeking minimum of the negative of the same function
* Univariate/Multivariate optimization (Taxonomy)
    Multivariable optimization problems
    1. Uncontrained opt
    2. Constrained opt
       * Equality/Inequality constraints
       
**Unconstrained optimization**

Example:

\begin{equation}
Objective function = f(x) = x^3 - 12x + 1
\end{equation}

\begin{equation}
Gradient function = f'(x) = 3x^2 - 12
\end{equation}

\begin{equation}
\Delta gradient = f"(x) = 6x
\end{equation}

* Turning points 
    * Locate turning points in the gradient function (first derivative)
    * Find x such that f(x) = 0
    * Can solve analytically via closed form or use the quadratic formula
    * Can solve numerically via gradient descent
    * Newton's Method (numerical and analytical gradient)
        * approximates f(x) == 0
        * Uses f'(x) and f"(x)
        * Use f"(x) to dtermine if candidate is max or min      
        
 * Iterative methods (Newton)
     * Start with x as candidate root
     * Generate sequence of xn-1, xn, xn+1 which converges into solution
     1. Initial guess
     2. Approxmiate f(x) by tangent at (xi,f(xi)
     3. Find where f_tangement_x0 = 0
     4. Repeat until convergence
     
Example:
\begin{equation}
f(x) = x^3 + 2x -4
\end{equation}
   
\begin{equation}
f"(x) = 3x^2 + 2
\end{equation}

Assuming 1 is first approximation of root, do one iteration of Newton-Raphson.

\begin{equation}
f(x) = 1^3 + 2(1) -4 = -1
\end{equation}

\begin{equation}
f'(x) = 3(1)^2 + 2 = 5
\end{equation}

\begin{equation}
x_2 = x_i - \frac{f(x_i)}{f'(x_i)} = 1 - \frac{-1}{5} = 1.2
\end{equation}

Convergence depends on good initial guess

**Gradient Descent**

Goal: Minimize f(x)

Univariate case: 

\begin{equation}
x^{i+1} = x^i - \alpha^i f'(x^i)
\end{equation}

Multivariate case: 

\begin{equation}
x^{i+1} = x^i - \alpha^i J'(x^i)
\end{equation}

* Learning rates
    * Fixed rate (with stochastic gradient descent)
    * Dynamic, decreasing learning rate to 0 so converges to global minimum
    * Or can be calculated
    
* Line Search
    * A procedure that chooses alpha long a line
    * f is minimized where gradient is othogonal to the search line
 

* Second Order Necessary Condition: Necessary
    * if f'(x) = 0, then it's necessary that f(x) is locally convex

* First Order Condition and Second Order Necessary Condition are NOT Sufficient
    * Doesn't distinguish btwn local mins and maxes or points of inflection
    
* Second Order Sufficient Condition
    * If FOC and SOSC: f"(x*) > 0, then x* is a local minimizer
    
**Unconstrained multivariate optimization**

1. FOC: Gradient function is 0
2. SOC: Check via Hessian is positive definite. Sufficient condition of extreme point to be a minimum is having a positive definite Hessian at such point.

**Convexity**

IFF every pair of points within the object and every point on a straight line is also in the object.

**GD & ML**

Use GD to find minimums of whatever objective function you are using (like MSE)

* Linear Regression
    * Minimize ||Xw-y||^2
    
* Classification
    * Minimize log(1+exp(-yx^Tw))
    
## Module 4

**Linear Regression Basics**

* Supervised learning
* Dependence of y on x (inputs) is linear
* Model is: Y = B0 + B1*X + irreducible error
* Predict yhat = B0_hat + B1_hat*x
* Goal: Minimize Residual Squared Errors (y-yhat)^2

**Assessing Accuracy**

* Standard error reflects how the estimator varies under repeated sampling
* Confidence intervals: give the probability the range contains a true value of the parameter
    * AKA a 95% change the interval contains the true value
* Hypothesis Testing
    * H0: No relationship btwn X and Y (B0 = 0)
    * H1: There's a relationship btwn X and Y (B0 != 0)
* R-squared (variance explained) = 1-RSS/TSS
* TSS == total summed squares == np.sum(np.square(y-yhat))

**Multiple Linear Regression**
* Y = B0 + B1*x1 + B2*x2...+BP*xp + error
* Best case
    * Predictors are uncorrelated (balanced design)
    * Each coefficient is tested separately
    * Freeze all other constants while testing is possible
* Issues
    * Correlations in predictors cause issues in interpretations and high variance
    * Claims of causality should be avoided 
* Forward Selection
    * Begin with null model that has an intercept but no predictors
    * Add the variables with the lowest RSS's
    * Continue until a stopping rule is satisifed, like a p-value hitting a thershold
* Backward Selection
    * All variables are on the model
    * Remove variable with largest p-value (least statistically sigificant)
    * Continue until stopping threshold is reached

This is the loss function aka Objective Function aka MSE: 

\begin{equation}\tag{1.3}
f(\boldsymbol{\theta}) = \frac{1}{m}\sum_{i=1}^{m}\left[ \boldsymbol{\theta}\cdot\mathbf{x}'_i - y_i\right]^2
\end{equation}
*The multivariate case is denoted with a J rather than an F*

Loss function is also written as:

\begin{equation}\tag{1.3}
LOSS_{MSE}(\boldsymbol{\theta}) = \frac{1}{n}\sum_{i=1}^{n}\left[ \boldsymbol{\theta}\cdot\mathbf{x}'_i - y_i\right]^2
\end{equation}

This is the formula used to compute the gradient (univariate):

\begin{equation}\tag{3.1}
\nabla_{ LOSS_{MSE}}(\boldsymbol{\theta}) = \frac{2}{n}\,\sum_{i=1}^{n}\left[ \boldsymbol{\theta}\cdot\mathbf{x}'_i - y_i\right] \cdot \mathbf{x}'
\end{equation}

This is the closed form solution aka Normal Equation:

\begin{equation}
\hat{\boldsymbol{\theta}} = (X^T\cdot{X})^{-1}\cdot{X}^{T}\cdot{y}
\end{equation}

Not the best idea due to computational challenges for large data-sets

**Breakdown of parameters**

\begin{equation}\boldsymbol{\theta}\end{equation}
The model's *parameter vector* where the bias term is the 0 index value. The rest are feature weights.

\begin{equation}\boldsymbol{\theta}^{T} \end{equation}
Same as what was said before, but transpose it.

\begin{equation}\boldsymbol{x}\end{equation}
Feature vector

\begin{equation}\boldsymbol{\theta}^{T}\cdot{x} \end{equation}
Dot product of the weights and input features...aka yhat == predictions

**Numerical solution**
\begin{equation}
lim\frac{f(x+h)-f(x)}{h}
\end{equation}

**How to compute the gradient**
1. Initialize the parameters $\theta$.
2. Compute the gradient $\nabla_{\boldsymbol{\theta}} f(\boldsymbol{\theta})$.
3. Update the parameters: $\theta_{\text{new}} = \theta_{\text{old}} - \eta \cdot \nabla_{\boldsymbol{\theta}} f(\boldsymbol{\theta}) $  

**Feature Selection**

The process of choosing relevant features for our model. 

Why we do feature selection:
* simplification of models to make them easier to interpret by researchers/users,
* shorter training times,
* to avoid the curse of dimensionality,
* enhanced generalization by reducing overfitting (formally, reduction of variance).

*Select K Best (features)*

Select features according to the k highest scores. It takes as a parameter a score function, which must be applicable to a pair (X, y). The score function must return an array of scores, one for each feature X[:,i] of X. Small pvalue means the feature is independent of y. Large pvalue means it's non-randomly related to y, so it maybe important. 

**More formulas**

$
\hat{y} = h_{\mathbf{\theta}}(\mathbf{x}) = \mathbf{\theta}^T \cdot \mathbf{x}
$


**MSE cost function for a Linear Regression model**

$
\text{MSE}(\mathbf{X}, h_{\mathbf{\theta}}) = \dfrac{1}{m} \sum\limits_{i=1}^{m}{(\mathbf{\theta}^T \cdot \mathbf{x}^{(i)} - y^{(i)})^2}
$


**Normal Equation**

$
\hat{\mathbf{\theta}} = (\mathbf{X}^T \cdot \mathbf{X})^{-1} \cdot \mathbf{X}^T \cdot \mathbf{y}
$


**Partial derivatives notation :**

$\frac{\partial}{\partial \theta_j} \text{MSE}(\mathbf{\theta})$


**Partial derivatives of the cost function**

$
\dfrac{\partial}{\partial \theta_j} \text{MSE}(\mathbf{\theta}) = \dfrac{2}{m}\sum\limits_{i=1}^{m}(\mathbf{\theta}^T \cdot \mathbf{x}^{(i)} - y^{(i)})\, x_j^{(i)}
$


**Gradient vector of the cost function**

$
\nabla_{\mathbf{\theta}}\, \text{MSE}(\mathbf{\theta}) =
\begin{pmatrix}
 \frac{\partial}{\partial \theta_0} \text{MSE}(\mathbf{\theta}) \\
 \frac{\partial}{\partial \theta_1} \text{MSE}(\mathbf{\theta}) \\
 \vdots \\
 \frac{\partial}{\partial \theta_n} \text{MSE}(\mathbf{\theta})
\end{pmatrix}
 = \dfrac{2}{m} \mathbf{X}^T \cdot (\mathbf{X} \cdot \mathbf{\theta} - \mathbf{y})
$


**Gradient Descent step**

$
\mathbf{\theta}^{(\text{next step})} = \mathbf{\theta} - \eta \nabla_{\mathbf{\theta}}\, \text{MSE}(\mathbf{\theta})
$


$$
J(\mathbf{\theta}) = \text{MSE}(\mathbf{\theta}) + \alpha \sum\limits_{i=1}^{n}\left| \theta_i \right|
$$


**Elastic Net cost function**

$
J(\mathbf{\theta}) = \text{MSE}(\mathbf{\theta}) + r \alpha \sum\limits_{i=1}^{m}\left| \theta_i \right| + \dfrac{1 - r}{2} \alpha \sum\limits_{i=1}^{m}{\theta_i^2}
$


** Lasso Regression cost function**

$$
J(\mathbf{\theta}) = \text{MSE}(\mathbf{\theta}) + \alpha \sum\limits_{i=1}^{n}\left| \theta_i \right|
$$

**Lasso Regression subgradient vector**

$$
g(\mathbf{\theta}, J) = \nabla_{\mathbf{\theta}}\, \text{MSE}(\mathbf{\theta}) + \alpha
\begin{pmatrix}
  \operatorname{sign}(\theta_1) \\
  \operatorname{sign}(\theta_2) \\
  \vdots \\
  \operatorname{sign}(\theta_n) \\
\end{pmatrix} \quad \text{where } \operatorname{sign}(\theta_i) =
\begin{cases}
-1 & \text{if } \theta_i < 0 \\
0 & \text{if } \theta_i = 0 \\
+1 & \text{if } \theta_i > 0
\end{cases}
$$


**Gradient Descent step**

$
\mathbf{\theta}^{(\text{next step})} = \mathbf{\theta} - \eta \nabla_{\mathbf{\theta}}\, \text{MSE}(\mathbf{\theta})
$

**Flavors of gradient descent**
* Stochastic
    * Do gradient descent learning one example at a time (i.e., update the weight one training example at a time).
    